### PR TITLE
fix(plugin-meetings): Fix timestamp

### PR DESCRIPTION
**Pull request by shnaaz**
# COMPLETES #< INSERT LINK TO ISSUE >
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-484765
## This pull request addresses

Issue is that if client's clock is wrong then we calculate the timeLeft for permission token to expire incorrectly hence refreshing the token when not required.
To work around this we now get the current time when we receive the token, we also extract iat (issued time) from permission token payload and workout the shift in time. We then use this to work out the adjusted current time before calculating the timeLeft for permission token to expire. This fix stops us from  refreshing the token when it' not required.

< DESCRIBE THE CONTEXT OF THE ISSUE >

## by making the following changes

< DESCRIBE YOUR CHANGES >

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [ ] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 1]
```diff
@@ -7979,16 +7981,20 @@ export default class Meeting extends StatelessWebexPlugin {
...
+    // adjusted time is calculated in case your machine time is wrong
+    const adjustedCurrentTime = currentTime - shiftInTime;
 
-    // substract current time from the permissionTokenExp
-    // (permissionTokenExp is a epoch timestamp, not a time to live duration)
-    const timeLeft = (permissionTokenExpValue - now) / 1000;
+    const timeLeft = (permissionTokenExpValue - adjustedCurrentTime) / 1000;
 
-    return {timeLeft, expiryTime: permissionTokenExpValue, currentTime: now};
+    return {timeLeft, expiryTime: permissionTokenExpValue, currentTime};
```
Comment (maxinteger) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:7997]: issue: if we adjust the `timeLeft` then we should do the same for `expiryTime`, so `expiryTime` should be `permissionTokenExpValue - shiftInTime`

**Review by maxinteger — CHANGES_REQUESTED** [review]

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 1 above)
Comment (shnaaz) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:7997]: No, expiryTime is being received in permissionToken payload, we are not adjusting timeLeft we are adjusting the current local time

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 2]
```diff
@@ -3540,6 +3541,7 @@ export default class Meeting extends StatelessWebexPlugin {
    */
   public setPermissionTokenPayload(permissionToken: string) {
     this.permissionTokenPayload = jwt.decode(permissionToken);
+    this.tokenReceivedTime = new Date().getTime();
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3544]: it would be good to group the `permissionTokenPayload` and `tokenReceivedTime` under some common property in the Meeting object like "permissionTokenInfo" or something like that, otherwise we end up with a lot of scattered properties in the Meeting object, especially that the name "tokenReceivedTime" doesn't say which token it is about

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 3]
```diff
@@ -7979,16 +7981,20 @@ export default class Meeting extends StatelessWebexPlugin {
     }
 
     const permissionTokenExpValue = Number(this.permissionTokenPayload.exp);
+    const permissionTokenIssuedTime = Number(this.permissionTokenPayload.iat);
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:7984]: maybe it would be worth renaming the variables that store values read from the permission token (iat and exp) so that it's clear that these are server time values not local client time ones

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 2 above)
Comment (shnaaz) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3544]: `tokenReceivedTime` can be renamed to `permissionTokenReceivedLocalTime`, if I remove `permissionTokenPalyload` from the meeting object and move it into another property say 'permissionTokenInfo' as suggested then that will have lot more code/unittest changes. Or we can have `permissionTokenPalyload` in 2 places, directly on meetingObject and also in the new property

Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3544]: > then that will have lot more code/unittest changes

but it's just a simple renaming everywhere from `.permissionTokenPayload` to `.permissionTokenInfo.payload`. Anyway, just renaming it to `permissionTokenReceivedLocalTime` is I guess fine too

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 4]
```diff
@@ -9914,18 +9929,58 @@ describe('plugin-meetings', () => {
...
+      meeting.permissionTokenReceivedLocalTime = now;
       assert.deepEqual(meeting.getPermissionTokenExpiryInfo(), {timeLeft: -1, expiryTime: Number(expiryTime), currentTime: now});
     });
+
+    describe('#getPermissionTokenExpiryInfo with wrong current Time', () => {
+      let now;
+      let clock;
+      beforeEach(() => {
+        // current time is 3 hours off
+        now = Date.now() + 10800000;
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:9950]: might be worth also adding a test where shiftTime is negative

**Review by marcin-bazyl — APPROVED** [review]

**Review by maxinteger — APPROVED** [review]
